[MS-68] Socket & Stomp & Redis 를 활용한 채팅기능 구현

tjdgns8439 opened the pull request:
## 요약 🎀

- WebSocket, Stopm, Redis pub/sub 활용 채팅 구현
- 주변 API 구현

## 상세 내용 🌈
### WebSocket, Stopm, Redis pub/sub 활용 채팅 구현

(최대한 그림과 함께 쉽게 보실 수 있도록 작성해보겠습니다.)
**(공약)제가 해당 기술스택을 선택한 이유는 다음 주 중으로 모택 블로그에 포스팅 진행하도록 하겠습니다**

**대략적인 중요 로직은 다음과 같이 동작합니다.**
![스크린샷 2024-05-05 오후 10 41 54](https://github.com/MODU-TAXI/server-api/assets/100510247/2f661a76-675d-4b85-8ab2-4e2715befde0)
1. 프론트에서 웹소켓 연결(jwt토큰 유효성 체크)
2. /sub/{roomId} 을 통해 채팅방 구독
2-2. StompHandler에서 매핑 정보 확인
2-3. 처음 들어오는 경우 방 참가 메세지 전달(이 때까지는 단일 Topic으로 전달합니다.)
2-4. RedisSubscriber가 캐치해서 실제 방에 방메 메세지 전달
> RedisSubscriber의 sendMessage로직은 RedisConfig에서 MessageListenerAdapter로 등록되어있습니다.
3. pub/chat 을 통한 메세지 전송 (pub으로 발행된 요청을 StompHandler를 거치지 않습니다.)
3-1. @MessageMapping("/chat")이 적용되어 있는 ChatController의 sendMessage 로직 수행
3-2. 토큰 유효성 검사 후 redis로 메세지 발행(publish)

---
여기까지만 이해하시면 전반적인 흐름은 이해 완료입니다!
---

이제 실제 제대로 동작하기 위한 Rest API를 구현해야 합니다.
대표적으로 방참가 가능여부 API, 방대기열 조회 API, 방참가자 조회 API, 방 퇴장 API 등이 있습니다.

**하지만 역시 가장 중요한 방 퇴장 API 에 대해 보다 구체적으로 설명드리겠습니다.**

Stomp는 프론트에서 웹소켓을 통해 보낸 몇 가지 요청에 대한 기능을 구현할 수 있습니다.
현재 저희가 캐치해서 상세 구현하는 기능은 다음과 같습니다.
- 웹소켓 연결 요청(StompCommand.CONNECT)
- 구독 요청(StompCommand.SUBSCRIBE)
- 웹소켓 종료 요청(StompCommand.DISCONNECT)
이 세가지만 보면 떠오를 수 있는 질문이 있습니다.
**"근데 구독이 있으면 구독을 끊어야 하는 요청이 있어야 하지 않나요?"**
기존에는 DISCONNECT 시 방 구독을 끊었습니다.
하지만 다음과 같은 문제가 발생하였습니다.
1. 앱 종료시 끊기는 웹소켓, 이에 따라 기존 매핑 관계도 끊겨버리는 이슈
2. 모종의 이유로 끊긴 웹소켓에도 연결관계가 끊겨버림.

따라서 구독 취소요청을 따로 진행해야겠다는 생각을 했습니다.
(_구독 취소 요청(StompCommand.UNSUBSCRIBE) 있기는 하지만.. 이 부분은 프론트 측의 반발로 무산되었습니다._)

**결론적으로!!** 프론트와 합의해 해당 요청을 Rest API로 구현하기로 하였습니다.
- 이에 따라 세션으로 방 매핑 정보를 저장했던 이전과 달리 세션과 멤버정보, 멤버정보와 방정보를 매핑하였습니다.
- 또한 레디스에 해당 매핑 정보를 저장하여 구독취소가 아닌 Disconnect일 때는 그냥 연결만 끊기게 하였고,
- 실제 구독 취소일 때 해당 멤버에 대한 방 매핑 정보를 삭제하였습니다.
- 이에 따라 방 인원수 감소, 퇴장 메세지 전송, 방 매핑 정보 삭제 등이 아래 코드에서 실행되게 되었습니다.
```java
deleteChatRoomInfo.java
--- 생략 ---
int count = ChatNickName.valueOf(chatRoomMappingInfo.getNickname()).getValue();
        // 클라이언트 퇴장 메시지 발송한다.
        ChatMessageRequestDto chatMessageRequestDto = new ChatMessageRequestDto(Long.valueOf(
                chatRoomMappingInfo.getRoomId()), MessageType.LEAVE, "",
                chatRoomMappingInfo.getNickname(), member.getId().toString(), LocalDateTime.now());

        sendChatMessage(chatMessageRequestDto);

        chatRoomRepository.removeUserByMemberIdEnterInfo(member.getId().toString());
        chatRoomRepository.minusUserCount(chatRoomMappingInfo.getRoomId(), count);
--- 생략 ---
```

여기까지의 흐름을 아신다면 다음 수정이 필요한 부가 기능들에 대해서는 언급만 하겠습니다.

### StompHanldler 내 함수
```java
// 비트마스킹으로 부여할 닉네임 선정
private String setNickName(String roomId){
        Long count = chatRoomRepository.getUserCount(roomId);
        int emptyBitIndex = findEmptyBitIndex(count);

        if (emptyBitIndex != -1) {
            // 해당하는 비트 인덱스에 맞는 ChatNickName 할당
            return ChatNickName.values()[emptyBitIndex].name();
        } else {
            // 더 이상 할당 가능한 닉네임이 없을 때의 처리
            return "nickname";
        }
    }

    // 비어있는 비트 인덱스 찾는 메소드
    private int findEmptyBitIndex(Long count);
```
### Rest API
**domain.chat.service.ChatService.java**
```java
deleteChatRoomInfo -> 방 매핑 정보 삭제
chatRoomEnterPossibel -> 특정 방에 들어갈 수 있는 지
getChatRoomInfo -> 멤버의 방 매핑 정보 조회
``` 

**domain.chatmessage.service.ChatMessageService.java**
```java
chatMessageResponseDtoList -> 채팅 메세지 리스트 조회
deleteChatMessage -> 특정 . 채팅 메세지 전부 삭제
``` 

**domain.roomwaiting.RoomWaitingService.java**
```java
applyForParticipate -> 방 참가 신청 (대기열에 참가)
acceptForParticipate -> 방 대기열에 있는 특정 멤버 수락
getParticipateInRoom -> 방에 있는 참가자 조회
getWaitingList -> 리스트에 있는 참가자 조회
``` 

## 질문 및 이외 사항 🚩
- 백로그에 대해 명시하겠습니다.
- [ ] 스웨거 반환값 정리
- [ ] Restful한 엔드포인트 정리
- [ ] 스톰프 에러 핸들러 작성
- [ ] 프론트와 기능 별로 맞춰가며 묶기(ex. 방 참가는 소켓 연결 후에 해야한다 등)

## 리뷰어에게 ✨
- 면접 준비 이슈로 많은 똥을 뿌리고 도망온 것 같은 느낌이 들어 굉장히 죄송스럽습니다.
- 며칠 기다려줄 수 없는 각자의 사정이 있으니 이틀만 부탁드리겠습니다.. 죄송합니다..

tjdgns8439 committed:
[MS-68] Feat: WebSocket 의존성 주입

tjdgns8439 committed:
[MS-68] Feat: 웹소켓&스톰프&레디스 세팅

tjdgns8439 committed:
[MS-68] Feat: 채팅메세지 구현

tjdgns8439 committed:
[MS-68] Feat: 채널 생성,구독 Redis 기능 구현

tjdgns8439 committed:
[MS-68] Feat: StompHandler 인터셉터 설정

tjdgns8439 committed:
[MS-68] Feat: StompHandler 작성

tjdgns8439 committed:
[MS-68] Feat: 메세지 핸들링 고도화

tjdgns8439 committed:
[MS-68] Chore: Todo 작성 - 인원수제한

tjdgns8439 committed:
[MS-68] Feat: 구독요청(채팅방 참가) 시 메세지 전송

YooJisu826 committed:
[MS-65] 설문조사 API (#40)

* [MS-65] Feat: Onboarding 관련 entity, enum 정의

* [MS-65] Feat: 설문조사 API

* [MS-65] Feat: 서버 최초 실행 시 설문조사 질문 자동 등록

tjdgns8439 committed:
[MS-68] Refactor: pub시 토큰을 헤더 대신 requestBody에 담기(임시)

tjdgns8439 committed:
[MS-98] 방 생성 예외처리 (#39)

* [MS-98] Feat: 방생성 시 잘못된 요청에 대한 예외처리

* [MS-98] Refactor: 위도, 경도 상수화

* [MS-98] Refactor: RoomRequstDto의 Description 필드 삭제

* [MS-100] Refactor: 요청, 응답 시 longitude, latitude 형식 통일

* [MS-100] Refactor: 요청, 응답 시 longitude, latitude 이름 수정

* [MS-100] Refactor: 업데이트 시 데이터 검증 로직 추가

jaewonLeeKOR committed:
[MS-91] 전화번호 메시지 인증 (#34)

* [MS-92] Refactor: 인증번호 생성 로직 분리

* 문자 인증 관련 에러코드 생성

* [MS-94] Refactor: 인증번호 관련 response dto 이름변경

* [MS-94] Refactor: service 클래스 인증번호 관련 response dto 변경

* [MS-95] Feat: 휴대전화 인증코드 발송 API

* [MS-96] 휴대전화 인증코드 검증 API

* [MS-101] Feat: sms인증 시 회원가입 key 검증 로직 추가

* config 서브모듈 동기화

jaewonLeeKOR committed:
[MS-102] 테스트용 서버 설정 (#41)

* [MS-103] Chore: 스웨거 버전 명시 방법 변경

* [MS-104] Feat: 테스트용 서버 배포 파이프라인 구축

* [MS-105] Fix: 사용하지 않는 엔드포인트 Hidden 처리

* [MS-102] 테스트서버 배포 파이프라인 도커 환경파일 경로 수정

* [MS-102] Fix: 테스트 배포 파이프라인 디렉토리 경로 수정

* [MS-102] chore: 테스트 서버용 스웨거 주소 변경

tjdgns8439 committed:
[MS-68] Feat: 임시 닉네임 부여 및 세션 저장방식 수정

tjdgns8439 committed:
[MS-68] Feat: Config 설정

tjdgns8439 committed:
[MS-68] Feat: 이미 방에 참가해있을 때 예외처리

tjdgns8439 committed:
[MS-51] Feat: 닉네임 부여

tjdgns8439 committed:
[MS-68] Fix: 인원수 업데이트 함수 변경

tjdgns8439 committed:
[MS-110] Feat: 채팅메세지 저장 및 조회

tjdgns8439 committed:
[MS-68] Chore: 디렉터리 구조 변경

tjdgns8439 committed:
[MS-68] Feat: pub/sub 구조 변경

tjdgns8439 committed:
[MS-112] Feat: 참여해있는 방 조회

tjdgns8439 committed:
[MS-109] Fix: 인원수 추가 로직 변경

- 기존에는 단순 disconnect를 경험한 재연결인지 퇴장 후 새로운 연결인 지 분간하지 않고 count를 올려줬습니다.
- 해당 문제를 매핑정보가 없을 때만 count하는 방식으로 수정해주었습니다.

tjdgns8439 committed:
[MS-68] Fix: 재참여시 메세지 전송하는 에러 제거

tjdgns8439 committed:
[MS-117] Feat: pub/sub시 메세지 dto 분리

tjdgns8439 committed:
[MS-116] Feat: 전송시간정보 추가

tjdgns8439 committed:
[MS-68] Feat: 존재하지 않는 방은 구독할 수 없게 처리

tjdgns8439 committed:
[MS-68] Feat: 메세지 pub 시 token 헤더에 담도록 수정

tjdgns8439 committed:
[MS-68] Feat: 방 매니저가 채팅방 퇴장하려 할 때 에러 반환

tjdgns8439 committed:
[MS-121] Feat: 대기열 참가, 수락 체크

jaewonLeeKOR committed:
[MS-123] Feat: 방참가, 채팅 정보, 채팅 메세지 API 반환 dto 수정

jaewonLeeKOR committed:
[MS-123] Fix: 누락된 방, 참가 에러 코드 추가

jaewonLeeKOR committed:
[MS-124] Fix: 방 세부 정보 조회 시 방장정보 추가

tjdgns8439 committed:
Merge branch 'develop' of github.com:MODU-TAXI/server-api into develop

tjdgns8439 committed:
Merge branch 'develop' into MS-68-setting-socket

tjdgns8439 committed:
[MS-68] Fix: 변경사항에 따른 에러해결

jaewonLeeKOR commented on src/main/java/com/modutaxi/api/common/config/websocket/StompHandler.java:
```diff
@@ -0,0 +1,152 @@
+    @Override
+    public Message<?> preSend(Message<?> message, MessageChannel channel) {
+        StompHeaderAccessor accessor = StompHeaderAccessor.wrap(message);
+        System.out.println("---------------------------------------");
+        System.out.println("Command: " + accessor.getCommand());
+        System.out.println("Destination: " + accessor.getDestination());
+        System.out.println("Message Headers: " + accessor.getMessageHeaders());
+        System.out.println("First Native Header: " + accessor.getFirstNativeHeader("token"));
+        System.out.println("Session ID: " + accessor.getSessionId());
+        System.out.println("User: " + accessor.getUser());
+        System.out.println("Subscription ID: " + accessor.getSubscriptionId());
```
운영 단계에서 필요한 로그인가요?

tjdgns8439 replied:
아닙니다. 하지만 제가 작업할 때 해당 로그들이 매우 도움되었습니다.
망고님께서 건네 받았을 때 로그가 남아있으면 좋을 것 같다고 생각하여 남겨두었습니다.

tjdgns8439 replied:
또한 프론트와 실시간으로 개발할 때 에러 핸들링을 해당 로그를 보면서 진행했습니다~ 
체크해놓고 완성되면 삭제하는 방향이 좋을 것 같습니다.

jaewonLeeKOR replied:
서비스 흐름과는 관련이 없는 부분이니까 메서드로 빼내는게 좋을것같습니다.
작업중이라도 혹시나 다른 사람이 맡아서 작업할수도 있는것이고 코드를 잘 정리해두는게 좋다고 생각합니다.

tjdgns8439 replied:
그럼 삭제하겠습니다!

jaewonLeeKOR commented on build.gradle:
```diff
@@ -80,6 +80,11 @@ dependencies {
+    // WebSocket
+    implementation 'org.springframework.boot:spring-boot-starter-websocket'
+
+    //Cool SMS
+    implementation 'net.nurigo:sdk:4.2.7'
     //Cool SMS
     implementation 'net.nurigo:sdk:4.2.7'
```
Depenency를 중복적으로 넣고 있어요.

tjdgns8439 replied:
그렇네요! 삭제하겠씁니다.

jaewonLeeKOR commented on src/main/java/com/modutaxi/api/common/config/websocket/StompHandler.java:
```diff
@@ -0,0 +1,152 @@
+                throw new BaseException(ChatErrorCode.FAULT_ROOM_ID);
+            }
+
+            //없는 방 연결하려 할 때 예외
+            Room room = roomRepository.findById(Long.valueOf(roomId)).orElseThrow(
+                    () -> new BaseException(RoomErrorCode.EMPTY_ROOM));
+
+            //이미 연결된 방이 있는데 애꿎은 방을 들어가려고 하면 컷
+            //연결되어 있는 방이 존재하면서 && 요청으로 들어온 roomId가 연결되어 있는 방과 다를 때
+            if (chatRoomMappingInfo != null && !roomId.equals(chatRoomMappingInfo.getRoomId())) {
+                throw new BaseException(ChatErrorCode.ALREADY_ROOM_IN);
+            }
+
+            String nickName = chatRoomMappingInfo != null ? chatRoomMappingInfo.getNickname() : setNickName(roomId);
+            int count = ChatNickName.valueOf(nickName).getValue();
+
+            if (chatRoomRepository.getUserCount(roomId) >= 15) {
+                throw new BaseException(ChatErrorCode.FULL_CHAT_ROOM);
+            }
+
+            if(chatRoomMappingInfo == null){
+                chatRoomMappingInfo = new ChatRoomMappingInfo(roomId, nickName);
+                chatRoomRepository.setUserEnterInfo(memberId, chatRoomMappingInfo);
+                // 채팅방의 인원수 체크
+                chatRoomRepository.plusUserCount(roomId, count);
+                chatService.sendChatMessage(new ChatMessageRequestDto(Long.valueOf(roomId),MessageType.JOIN,"",
+                        chatRoomMappingInfo.getNickname(),memberId, LocalDateTime.now()));
+            }
+        }
+
+        else if (StompCommand.DISCONNECT == accessor.getCommand()) {
+            String sessionId = (String) accessor.getSessionId();
+            // 세션에 대한 정보 삭제
+            chatRoomRepository.removeUserBySessionIdEnterInfo(sessionId);
+            //웹소켓은 끊기면 걍 끊긴거다~
+            //대신 방 정보 매핑 삭제는 따로 API 호출을 해주어야 하는 것~
+        }
+
+        return message;
+    }
```
매서드 하나가 너무 크네요. 조건과 단게에 따라 여러개로 나눌 수 있을거 같습니다.

tjdgns8439 replied:
네 맞습니다! 하지만 요청사항에 따른 수정사항이 워낙 많습니다.
메서드 분리는 완성이 되었을 때 하는 걸 선호하는 편입니다.
세부적인 메서드 수정이 너무 많아서 우선 진행하고 완성된 배포 시에는 매우 잘게 모듈화하려는 데 어떻게 생각하시나요?

jaewonLeeKOR commented on src/main/java/com/modutaxi/api/common/config/websocket/StompHandler.java:
```diff
@@ -0,0 +1,152 @@
+        }
+
+        else if (StompCommand.DISCONNECT == accessor.getCommand()) {
+            String sessionId = (String) accessor.getSessionId();
```
sessionId는 이미 String인데 String으로 명시하는 이유가 뭔가요

tjdgns8439 replied:
삭제 하겠습니다~

jaewonLeeKOR commented on src/main/java/com/modutaxi/api/common/config/websocket/StompHandler.java:
```diff
@@ -0,0 +1,152 @@
+            String sessionId = (String) accessor.getSessionId();
+            // 세션에 대한 정보 삭제
+            chatRoomRepository.removeUserBySessionIdEnterInfo(sessionId);
+            //웹소켓은 끊기면 걍 끊긴거다~
+            //대신 방 정보 매핑 삭제는 따로 API 호출을 해주어야 하는 것~
```
추후에 할 작업을 정리해둔건가요?

tjdgns8439 replied:
망고 님이 제가 없는 동안 넘겨받을 것이기에 이해를 돕기 위해 남겨놓은 주석입니다

jaewonLeeKOR commented on src/main/java/com/modutaxi/api/common/config/websocket/StompHandler.java:
```diff
@@ -0,0 +1,152 @@
+            // 해당하는 비트 인덱스에 맞는 ChatNickName 할당
+            return ChatNickName.values()[emptyBitIndex].name();
+        } else {
+            // 더 이상 할당 가능한 닉네임이 없을 때의 처리
+            return "nickname";
```
이부분은 그냥 이름을 모두 동일하게 `nickname`으로 만드는건가요?

tjdgns8439 replied:
사실 해당 닉네임까지 가는 일은 없습니다.
만약 그렇다면 에러를 호출해야 합니다.

tjdgns8439 replied:
근데 StompErrorHandler가 아직 구현이 안돼서 못하는군요

jaewonLeeKOR commented on src/main/java/com/modutaxi/api/common/exception/errorcode/ChatErrorCode.java:
```diff
@@ -0,0 +1,22 @@
+@Getter
+@RequiredArgsConstructor
+public enum ChatErrorCode implements ErrorCode {
+    FULL_CHAT_ROOM("CHAT_001", "방이 꽉 차있습니다", HttpStatus.BAD_REQUEST),
+    FAULT_ROOM_ID("CHAT_002", "잘못된 방", HttpStatus.BAD_REQUEST),
+    ALREADY_ROOM_IN("CHAT_003", "이미 방에 참가해있잖아~", HttpStatus.BAD_REQUEST),
+    ALREADY_ROOM_OUT("CHAT_004", "이미 나가있잖아 인마", HttpStatus.BAD_REQUEST),
```
`400` 번 에러는 프론트에서 직접 사용자에게 보여주는 에러로 알고 있습니다. 에러 메세지를 정리가 필요할거 같네요.

tjdgns8439 replied:
네 저도 그렇게생각합니다!

jaewonLeeKOR commented on src/main/java/com/modutaxi/api/domain/chatmessage/entity/ChatMessage.java:
```diff
@@ -0,0 +1,53 @@
+
+    private LocalDateTime dateTime;
+
+    public static ChatMessage toEntity(ChatMessageRequestDto messageRequestDto, Room room){
+        return ChatMessage.builder()
+                .room(room)
+                .messageType(messageRequestDto.getType())
+                .content(messageRequestDto.getContent())
+                .sender(messageRequestDto.getSender())
+                .memberId(messageRequestDto.getMemberId())
+                .dateTime(messageRequestDto.getDateTime())
+                .build();
```
이전에 협의한 대로 `ChatMessageMapper` 클래스의 `toEntity` 메서드로 변경하는것이 좋을거 같아요.

tjdgns8439 replied:
넵 변경하겠습니다!

jaewonLeeKOR commented on src/main/java/com/modutaxi/api/domain/roomwaiting/RoomWaitingService.java:
```diff
@@ -0,0 +1,124 @@
+    }
+
+    // TODO: 5/2/24 귀찮으니 코드 재사용
+    public MemberRoomInResponseList getParticipateInRoom(Long roomId){
+        Room room = roomRepository.findById(Long.valueOf(roomId)).orElseThrow(
+                () -> new BaseException(RoomErrorCode.EMPTY_ROOM));
+        Set<String> memberIdSet = chatRoomRepository.findRoomInList(roomId.toString());
+
+        List<Long> memberIdList = new ArrayList<>(memberIdSet).stream().map(Long::valueOf).toList();
+        List<Member> waitingUserList = getMemberService.getMemberList(memberIdList);
+
+        return new MemberRoomInResponseList(
+            waitingUserList.stream()
+                .map(MemberRoomInResponse::toDto)
+                .collect(Collectors.toList()));
+    }
+
+    public RoomWaitingResponseList getWaitingList(Long roomId){
+        Room room = roomRepository.findById(roomId).orElseThrow(
+                () -> new BaseException(RoomErrorCode.EMPTY_ROOM));
+
+        Set<String> memberIdSet = chatRoomRepository.findWaitingList(roomId.toString());
+
+        List<Long> memberIdList = new ArrayList<>(memberIdSet).stream().map(Long::valueOf).toList();
+        List<Member> waitingUserList = getMemberService.getMemberList(memberIdList);
+
+        return new RoomWaitingResponseList(
+            waitingUserList.stream()
+                .map(RoomWaitingResponse::toDto)
+                .collect(Collectors.toList()));
+    }
```
두 메서드는 한가지 조건을 제외하고는 완벽히 동일한 코드 같네요. 합치는게 좋을거 같습니다.

tjdgns8439 replied:
분리는 하되 공통된 내부 로직을 분리하는 편이 맞을 것 같습니다!

tjdgns8439 committed:
[MS-68] Refactor: 리뷰 반영

- 중복 의존성 주입 수정
- 에러 메세지 수정
- sessionId
- 의미없는 형변환 제거
- toEntity함수를 Mapper 클래스 내부로 이동

YooJisu826 reviewed (dismissed):
질문은 따로 달겠습니다!

jaewonLeeKOR reviewed (dismissed):
프론트 측에서 빠른병합을 계속해서 요구하는 상황이니,
이 이슈에 대해서는 따로 discussion을 생성해서 얘기하도록 하죠.

tjdgns8439 committed:
[MS-68] Chore: 로그 출력 삭제

tjdgns8439 committed:
Merge branch 'develop' into MS-68-setting-socket

jaewonLeeKOR reviewed (approved):
LGTM
